### DISCUSSION TITLE
Secondary development with respect to general contact interface, /inter/type24

### DISCUSSION THREAD

**Discussion by ZhengZQ007** [OpenRadioss/OpenRadioss#1731] [Q&A] [answered]
Recently, I have studied theory about computational contact mechanics, and I want to do some secondary development with respect to OpenRadioss's general contact interface, /inter/type24. I have browsed relevant source codes in OpenRadioss and tried to understand the algorithm behind. But it seems very difficult since I do not know the purpose of relevant subroutines and essential data structures. Could anyone give some help?

**Comment by qiangRadioss** [comment]
Hi Zheng,

thank you for your interest in OpenRadioss, especially for the contact, key subject and very complex to discuss.
Before going into the detail, could you tell us some context of your future development, what kind of application of it? just to be sure that type24 is the best choice for you ?

Regards.

Qiang.

**Reply by ZhengZQ007** [reply]
Hi Qiang,

Thanks for your kindly suggestion. I am interested in type 24 because it is a mature general nodes-to-surface contact interface. It covers the most features of  computational contact mechanics. Based on type24, I could make a research on geometrically exact theory for contact between arbitrary shaped bodies in the future. Could you give me some advice on how to understand the algorithm behind type24? 

Best Regards.

Zheng.

**Reply by qiangRadioss** [reply, marked as answer]
Hi Zheng,

Right, type 24 is N/S (Node-to-Surface) contact (edge option is also N/S), especially designed for solid model and now mature in the applications of electronic industry. If your applications are rather crashworthiness with shell&solid model, i propose you to look at type 25 (developed after type24 and have the true edge options), or type 7 (N/S as well) which gives high depenetration quality and robustness.
A few words for type24 :
. formulation : intersection +sliding
. mono-impact of each secondary node
. main segment (surface) : semi nodal normal+interpolation
. 2 segments for each shell (upper&lower)
I propose you to look at first the parts in Starter of type 24 : initialization +initial contact detection
driven subroutine : 
Starter :  source\interfaces\inter3d1\inint3.F
Engine :  source\interfaces\intsort\inttri.F (sorting); source\interfaces\int24\i24main.F (contact force)

hope this helps

Qiang.

**Reply by ZhengZQ007** [reply]
Hi Qiang,

Thanks very much for your detailed explanations on type 24 about its features and limitations. Based on your information, I can choose an appropriate contact type in future modellings. Sincere thanks also for your helpful instructions on how to read the source codes about type 24, which is very helpful for me. By the way, your explanation on the calculation of rotation matrix [R] for HEPH brick element last time helps me a lot. 

Best Regards.

Zheng

**Reply by ZhengZQ007** [reply]
Hi Qiang,

For type24, there are fictive nodes added for external border edges to consider edge-to-edge contact. For these fictive nodes, how their contact forces to be scattered to global nodal force vector?  It seems that the contact force of a certain fictive node uniformly spreads to real nodes in the same surface. 

Best Regards.

Zheng

**Reply by qiangRadioss** [reply]
Hi Zheng,

yes, the contact forces on the fictive nodes have been condensed to their edge nodes. (see I24FOR1_FIC)

Regards.

Qiang.

**Reply by ZhengZQ007** [reply]
Thanks very much.

**Comment by QuatorzeM** [comment]
Seems interesting. Bon courage!
